File: src/main/java/io/xdea/xmux/forum/controller/ThreadController.java
package io.xdea.xmux.forum.controller;

import io.xdea.xmux.forum.dto.CommonGrpcApi;
import io.xdea.xmux.forum.dto.SavedGrpcApi;
import io.xdea.xmux.forum.dto.ThreadGrpcApi;
import io.xdea.xmux.forum.interceptor.AuthInterceptor;
import io.xdea.xmux.forum.model.Forum;
import io.xdea.xmux.forum.model.Thread;
import io.xdea.xmux.forum.model.ThreadWithInfo;
import io.xdea.xmux.forum.service.AliyunGreenService;
import io.xdea.xmux.forum.service.ForumService;
import io.xdea.xmux.forum.service.NotifService;
import io.xdea.xmux.forum.service.ThreadService;
import com.google.protobuf.Empty;
import com.google.protobuf.Timestamp;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;

import org.apache.commons.lang3.StringUtils;

public abstract class ThreadController extends NotifController {

    protected final ThreadService threadService;

    protected ThreadController(ForumService forumService, NotifService notifService, ThreadService threadService, AliyunGreenService aliyunGreenService) {
        super(forumService, notifService, aliyunGreenService);
        this.threadService = threadService;
    }

    private ThreadGrpcApi.Thread buildThread(ThreadWithInfo thread) {
        var builder = ThreadGrpcApi.Thread.newBuilder()
                .setId(thread.getId())
                .setCreateAt(Timestamp.newBuilder()
                        .setSeconds(thread.getCreateAt().getTime() / 1000))
                .setUpdateAt(Timestamp.newBuilder()
                        .setSeconds(thread.getUpdateAt().getTime() / 1000))
                .setLastUpdate(Timestamp.newBuilder()
                        .setSeconds(thread.getLastUpdate().getTime() / 1000))
                .setDigest(thread.getDigest())
                .setLikes(thread.getLikes())
                .setPinned(thread.getPinned())
                .setTitle(thread.getTitle())
                .setUid(thread.getUid())
                .setLiked(thread.getLiked())
                .setSaved(thread.getSaved())
                .setPosts(thread.getPosts());

        var bodyCase = ThreadGrpcApi.Thread.BodyCase
                .forNumber(thread.getBodyType());
        if (bodyCase.equals(ThreadGrpcApi.Thread.BodyCase.PLAINCONTENT)) {
            var contentBuilder = CommonGrpcApi.PlainContent.newBuilder();
            contentBuilder.setContent((String) thread.getBody().get("content"));
            for (String img : (List<String>) thread.getBody().get("images")) {
                contentBuilder.addImages(CommonGrpcApi.Image.newBuilder().setUrl(img).build());
            }

            builder.setPlainContent(contentBuilder.build());

        } else if (bodyCase.equals(ThreadGrpcApi.Thread.BodyCase.MARKDOWNCONTENT)) {
            builder.setMarkdownContent(CommonGrpcApi.MarkdownContent.newBuilder()
                    .setContent((String) thread.getBody().get("content")).build());
        }

        return builder.build();
    }

    @Override
    public void createThread(ThreadGrpcApi.CreateThreadReq request, StreamObserver<ThreadGrpcApi.CreateThreadResp> responseObserver) {
        String uid = AuthInterceptor.getUid();
        
        // Check if the content is sensitive
        if (request.getBodyCase().equals(ThreadGrpcApi.CreateThreadReq.BodyCase.PLAINCONTENT)) {
            if (!aliyunGreenService.checkText(request.getTitle() + " - " + request.getPlainContent().getContent(), uid,
                    AliyunGreenService.ContentType.THREAD)) {
                responseObserver.onError(Status.INVALID_ARGUMENT
                    .withDescription("error.censored_content")
                    .asException());
                return;
            }
        } else if (request.getBodyCase().equals(ThreadGrpcApi.CreateThreadReq.BodyCase.MARKDOWNCONTENT)) {
            if (!aliyunGreenService.checkText(request.getTitle() + " - " + request.getMarkdownContent().getContent(), uid,
                    AliyunGreenService.ContentType.THREAD)) {
                responseObserver.onError(Status.INVALID_ARGUMENT
                    .withDescription("error.censored_content")
                    .asException());
                return;
            }
        }

        Date nowTime = new Date();
        Thread thread = new Thread()
                .withTitle(request.getTitle())
                .withForumId(request.getForumId())
                .withUid(uid)
                .withCreateAt(nowTime)
                .withLastUpdate(nowTime)
                .withUpdateAt(nowTime)
                .withLikes(0)
                .withPosts(0)
                .withDigest(false)
                .withPinned(false);

        thread.setBodyType(request.getBodyCase().getNumber());
        // Store content according to type
        HashMap<String, Object> content = new HashMap<>();
        if (request.getBodyCase().equals(ThreadGrpcApi.CreateThreadReq.BodyCase.PLAINCONTENT)) {
            content.put("content", request.getPlainContent().getContent());
            // Store image URLs
            ArrayList<String> imgList = new ArrayList<>();
            for (var image : request.getPlainContent().getImagesList()) {
                imgList.add(image.getUrl());
            }
            content.put("images", imgList);

        } else if (request.getBodyCase().equals(ThreadGrpcApi.CreateThreadReq.BodyCase.MARKDOWNCONTENT)) {
            content.put("content", request.getMarkdownContent().getContent());
        }
        thread.setBody(content);

        if (!threadService.create(thread))
            throw new RuntimeException("threadService.create returned false");

        var resp = ThreadGrpcApi.CreateThreadResp
                .newBuilder().setThreadId(thread.getId()).build();
        responseObserver.onNext(resp);
        responseObserver.onCompleted();
    }

    @Override
    public void removeThread(ThreadGrpcApi.RemoveThreadReq request, StreamObserver<Empty> responseObserver) {
        String uid = AuthInterceptor.getUid();
        Thread thread = threadService.getById(request.getThreadId());
        // Check if the user has privilege to remove the thread
        if (thread == null) {
            responseObserver.onError(Status.INVALID_ARGUMENT
                    .withDescription("Thread does not exist").asException());
            return;
        }
        String threadUid = thread.getUid();
        if (threadUid == null || !threadUid.equals(uid)) {
            responseObserver.onError(Status.PERMISSION_DENIED
                    .withDescription("Not the owner of the thread").asException());
            return;
        }

        // Check passed, soft remove the thread
        if (!threadService.softRemove(request.getThreadId()))
            throw new RuntimeException("threadService.softRemove returned false");

        responseObserver.onNext(Empty.getDefaultInstance());
        responseObserver.onCompleted();
    }

    @Override
    public void updateThread(ThreadGrpcApi.UpdateThreadReq request, StreamObserver<Empty> responseObserver) {
        if (StringUtils.isAnyBlank(request.getTitle(), request.getBody())) {
            responseObserver.onError(Status.INVALID_ARGUMENT
                    .withDescription("error.content_empty").asException());
            return;
        }
        if (!threadService.update(request.getId(), request.getTitle(), request.getBody()))
            throw new RuntimeException("threadService.update returned false");
        responseObserver.onNext(Empty.getDefaultInstance());
        responseObserver.onCompleted();
    }

    @Override
    public void getThreads(ThreadGrpcApi.GetThreadsReq request, StreamObserver<ThreadGrpcApi.GetThreadsResp> responseObserver) {
        String uid = AuthInterceptor.getUid();
        int forumId = request.getForumId();
        final var respBuilder = ThreadGrpcApi.GetThreadsResp.newBuilder();
        var threads = threadService.get(request.getOffset(), request.getCount(), forumId, uid, request.getOrderingValue());
        threads.forEach(thread ->
                respBuilder.addThreads(buildThread(thread))
        );

        responseObserver.onNext(respBuilder.build());
        responseObserver.onCompleted();
    }

    @Override
    public void likeThread(ThreadGrpcApi.LikeThreadReq request, StreamObserver<Empty> responseObserver) {
        String uid = AuthInterceptor.getUid();
        if (request.getLike() > 0) {
            if (threadService.upvote(request.getThreadId(), uid)) {
                // Like successfully, send notif
                Thread thread = threadService.getById(request.getThreadId());
                notifService.createThreadLikeNotif(thread.getId(), thread.getUid(), uid);
            }
        } else if (request.getLike() == 0) {
            threadService.cancelVote(request.getThreadId(), uid);
        } else {
            threadService.downvote(request.getThreadId(), uid);
        }
        responseObserver.onNext(Empty.getDefaultInstance());
        responseObserver.onCompleted();
    }

    @Override
    public void pinThread(ThreadGrpcApi.PinThreadReq request, StreamObserver<Empty> responseObserver) {
        String uid = AuthInterceptor.getUid();
        final Thread thread = threadService.getById(request.getThreadId());
        final Forum forum = forumService.getById(thread.getForumId());
        if (!forum.getCreatorUid().equals(uid)) {
            responseObserver.onError(Status.PERMISSION_DENIED.asException());
            return;
        }

        if (!threadService.togglePinned(request.getThreadId()))
            throw new RuntimeException("threadService.togglePinned returned false");
        responseObserver.onNext(Empty.getDefaultInstance());
        responseObserver.onCompleted();
    }

    @Override
    public void digestThread(ThreadGrpcApi.DigestThreadReq request, StreamObserver<Empty> responseObserver) {
        String uid = AuthInterceptor.getUid();
        final Thread thread = threadService.getById(request.getThreadId());
        final Forum forum = forumService.getById(thread.getForumId());
        if (!forum.getCreatorUid().equals(uid)) {
            responseObserver.onError(Status.PERMISSION_DENIED.asException());
            return;
        }

        if (!threadService.toggleDigest(request.getThreadId()))
            throw new RuntimeException("threadService.toggleDigest returned false");
        responseObserver.onNext(Empty.getDefaultInstance());
        responseObserver.onCompleted();
    }

    @Override
    public void getSavedThreads(SavedGrpcApi.GetSavedThreadsReq request, StreamObserver<ThreadGrpcApi.GetThreadsResp> responseObserver) {
        String uid = AuthInterceptor.getUid();
        final var saved = threadService.getSaved(request.getOffset(), request.getCount(), uid);
        final var builder = ThreadGrpcApi.GetThreadsResp.newBuilder();
        saved.forEach(thread -> {
            builder.addThreads(buildThread(thread));
        });
        responseObserver.onNext(builder.build());
        responseObserver.onCompleted();
    }
}

File: src/main/java/io/xdea/xmux/forum/controller/ThreadController.kt
package io.xdea.xmux.forum.controller

import com.google.protobuf.Empty
import com.google.protobuf.Timestamp
import io.grpc.Status
import io.grpc.stub.StreamObserver
import io.xdea.xmux.forum.dto.CommonGrpcApi
import io.xdea.xmux.forum.dto.CommonGrpcApi.PlainContent
import io.xdea.xmux.forum.dto.SavedGrpcApi.GetSavedThreadsReq
import io.xdea.xmux.forum.dto.ThreadGrpcApi
import io.xdea.xmux.forum.dto.ThreadGrpcApi.*
import io.xdea.xmux.forum.interceptor.AuthInterceptor
import io.xdea.xmux.forum.model.Thread
import io.xdea.xmux.forum.model.ThreadWithInfo
import io.xdea.xmux.forum.service.ForumService
import io.xdea.xmux.forum.service.NotifService
import io.xdea.xmux.forum.service.ThreadService
import java.util.*
import java.util.function.Consumer

abstract class ThreadController protected constructor(
    forumService: ForumService,
    notifService: NotifService,
    protected val threadService: ThreadService
) : NotifController(forumService, notifService) {
    private fun buildThread(thread: ThreadWithInfo): ThreadGrpcApi.Thread {
        val builder = ThreadGrpcApi.Thread.newBuilder()
            .setId(thread.id)
            .setCreateAt(
                Timestamp.newBuilder()
                    .setSeconds(thread.createAt.time / 1000)
            )
            .setUpdateAt(
                Timestamp.newBuilder()
                    .setSeconds(thread.updateAt.time / 1000)
            )
            .setLastUpdate(
                Timestamp.newBuilder()
                    .setSeconds(thread.lastUpdate.time / 1000)
            )
            .setDigest(thread.digest)
            .setLikes(thread.likes)
            .setPinned(thread.pinned)
            .setTitle(thread.title)
            .setUid(thread.uid)
            .setLiked(thread.liked)
            .setSaved(thread.saved)
            .setPosts(thread.posts)

        val bodyCase = ThreadGrpcApi.Thread.BodyCase
            .forNumber(thread.bodyType)
        if (bodyCase == ThreadGrpcApi.Thread.BodyCase.PLAINCONTENT) {
            val contentBuilder = PlainContent.newBuilder()
            contentBuilder.setContent(thread.body["content"] as String?)
            for (img in (thread.body["images"] as List<String>?)!!) {
                contentBuilder.addImages(CommonGrpcApi.Image.newBuilder().setUrl(img).build())
            }

            builder.setPlainContent(contentBuilder.build())
        } else if (bodyCase == ThreadGrpcApi.Thread.BodyCase.MARKDOWNCONTENT) {
            builder.setMarkdownContent(
                CommonGrpcApi.MarkdownContent.newBuilder()
                    .setContent(thread.body["content"] as String?).build()
            )
        }

        return builder.build()
    }

    override fun createThread(request: CreateThreadReq, responseObserver: StreamObserver<CreateThreadResp>) {
        val uid = AuthInterceptor.UID.get()
        val nowTime = Date()
        val thread = Thread()
            .withTitle(request.title)
            .withForumId(request.forumId)
            .withUid(uid)
            .withCreateAt(nowTime)
            .withLastUpdate(nowTime)
            .withUpdateAt(nowTime)
            .withLikes(0)
            .withPosts(0)
            .withDigest(false)
            .withPinned(false)

        thread.bodyType = request.bodyCase.number
        // Store content according to type
        val content = HashMap<String?, Any?>()
        if (request.bodyCase == CreateThreadReq.BodyCase.PLAINCONTENT) {
            content["content"] = request.plainContent.content
            // Store image URLs
            val imgList = ArrayList<String>()
            for (image in request.plainContent.imagesList) {
                imgList.add(image.url)
            }
            content["images"] = imgList
        } else if (request.bodyCase == CreateThreadReq.BodyCase.MARKDOWNCONTENT) {
            content["content"] = request.markdownContent.content
        }
        thread.body = content

        if (!threadService.create(thread)) throw RuntimeException("threadService.create returned false")

        val resp = CreateThreadResp
            .newBuilder().setThreadId(thread.id).build()
        responseObserver.onNext(resp)
        responseObserver.onCompleted()
    }

    override fun removeThread(request: RemoveThreadReq, responseObserver: StreamObserver<Empty>) {
        val uid = AuthInterceptor.UID.get()
        val thread = threadService.getById(request.threadId)
        // Check if the user has privilege to remove the thread
        if (thread == null) {
            responseObserver.onError(
                Status.INVALID_ARGUMENT
                    .withDescription("Thread does not exist").asException()
            )
            return
        }
        val threadUid = thread.uid
        if (threadUid == null || threadUid != uid) {
            responseObserver.onError(
                Status.PERMISSION_DENIED
                    .withDescription("Not the owner of the thread").asException()
            )
            return
        }

        // Check passed, remove thread and all its posts
        if (!threadService.hardRemove(request.threadId)) throw RuntimeException("threadService.hardRemove returned false")

        responseObserver.onNext(Empty.getDefaultInstance())
        responseObserver.onCompleted()
    }

    override fun updateThread(request: UpdateThreadReq, responseObserver: StreamObserver<Empty>) {
        if (!threadService.update(
                request.id,
                request.title,
                request.body
            )
        ) throw RuntimeException("threadService.update returned false")
        responseObserver.onNext(Empty.getDefaultInstance())
        responseObserver.onCompleted()
    }

    override fun getThreads(request: GetThreadsReq, responseObserver: StreamObserver<GetThreadsResp>) {
        val uid = AuthInterceptor.UID.get()
        val forumId = request.forumId
        val respBuilder = GetThreadsResp.newBuilder()
        val threads = threadService[request.offset, request.count, forumId, uid, request.orderingValue]
        threads.forEach(
            Consumer { thread: ThreadWithInfo -> respBuilder.addThreads(buildThread(thread)) }
        )

        responseObserver.onNext(respBuilder.build())
        responseObserver.onCompleted()
    }

    override fun likeThread(request: LikeThreadReq, responseObserver: StreamObserver<Empty>) {
        val uid = AuthInterceptor.UID.get()
        if (request.like > 0) {
            if (!threadService.upvote(
                    request.threadId,
                    uid
                )
            ) throw RuntimeException("threadService.upvote returned false")
        } else if (request.like == 0) {
            if (!threadService.cancelVote(
                    request.threadId,
                    uid
                )
            ) throw RuntimeException("threadService.cancleVote returned false")
        } else {
            if (!threadService.downvote(
                    request.threadId,
                    uid
                )
            ) throw RuntimeException("threadService.downvote returned false")
        }
        responseObserver.onNext(Empty.getDefaultInstance())
        responseObserver.onCompleted()
    }

    override fun pinThread(request: PinThreadReq, responseObserver: StreamObserver<Empty>) {
        val uid = AuthInterceptor.UID.get()
        val thread = threadService.getById(request.threadId)
        val forum = forumService.getById(thread.forumId)
        if (forum.creatorUid != uid) {
            responseObserver.onError(Status.PERMISSION_DENIED.asException())
            return
        }

        if (!threadService.togglePinned(request.threadId)) throw RuntimeException("threadService.togglePinned returned false")
        responseObserver.onNext(Empty.getDefaultInstance())
        responseObserver.onCompleted()
    }

    override fun digestThread(request: DigestThreadReq, responseObserver: StreamObserver<Empty>) {
        val uid = AuthInterceptor.UID.get()
        val thread = threadService.getById(request.threadId)
        val forum = forumService.getById(thread.forumId)
        if (forum.creatorUid != uid) {
            responseObserver.onError(Status.PERMISSION_DENIED.asException())
            return
        }

        if (!threadService.toggleDigest(request.threadId)) throw RuntimeException("threadService.toggleDigest returned false")
        responseObserver.onNext(Empty.getDefaultInstance())
        responseObserver.onCompleted()
    }

    override fun getSavedThreads(request: GetSavedThreadsReq, responseObserver: StreamObserver<GetThreadsResp>) {
        val uid = AuthInterceptor.UID.get()
        val saved = threadService.getSaved(request.offset, request.count, uid)
        val builder = GetThreadsResp.newBuilder()
        saved.forEach(Consumer { thread: ThreadWithInfo ->
            builder.addThreads(buildThread(thread))
        })
        responseObserver.onNext(builder.build())
        responseObserver.onCompleted()
    }
}
